core/domain/classifier_services_test.py

```diff
@@ -0,0 +1,166 @@
+
+    def test_model_save_load(self):
+        self.assertEquals(
+            self.string_classifier._doc_count,
+            len(self._EXAMPLES))
+        model = self.string_classifier.to_dict()
+        model['_doc_count'] = 9
+        self.assertEquals(model['_doc_count'], 9)
+        self.assertEquals(
+            self.string_classifier._doc_count,
+            len(self._EXAMPLES))
+        self.string_classifier.add_examples(self._NEW_EXAMPLES)
+        self.assertEquals(
+            self.string_classifier._doc_count,
+            len(self._EXAMPLES) + len(self._NEW_EXAMPLES))
+        self.assertEquals(model['_doc_count'], 9)
+        self.string_classifier.from_dict(model)
+        self.assertEquals(self.string_classifier._doc_count, 9)
+        self.assertEquals(model['_doc_count'], 9)
+
+    def test_get_word_id(self):
+        word_count = self.string_classifier._word_count
+        self.string_classifier._get_word_id('_non_existent_word_1')
+        self.assertEquals(self.string_classifier._word_count, word_count + 1)
+        self.string_classifier._get_word_id('i')
+        self.assertEquals(self.string_classifier._word_count, word_count + 1)
+        self.string_classifier._get_word_id('_non_existent_word_2')
+        self.assertEquals(self.string_classifier._word_count, word_count + 2)
+
+    def test_get_label_id(self):
+        label_count = self.string_classifier._label_count
+        self.string_classifier._get_label_id('_non_existent_label_1')
+        self.assertEquals(self.string_classifier._label_count, label_count + 1)
+        self.string_classifier._get_label_id('food')
+        self.assertEquals(self.string_classifier._label_count, label_count + 1)
+        self.string_classifier._get_label_id('_non_existent_label_2')
+        self.assertEquals(self.string_classifier._label_count, label_count + 2)
+
+    def test_label_begins_with_underscore(self):
+        self.string_classifier.add_examples([['example doc', ['good_label']]])
+        with self.assertRaises(Exception):
+            self.string_classifier.add_examples(
+                [['example doc', ['_bad_label']]])
+
+    def test_reload_valid_state(self):
+        self.string_classifier.load_examples(self._NEW_EXAMPLES)
+        self.assertEquals(self.string_classifier._label_count, 1)
+        self.assertEquals(
+            self.string_classifier._doc_count,
+            len(self._NEW_EXAMPLES))
+        self.assertEquals(self.string_classifier._word_count, 34)
+        self._validate_instance(self.string_classifier)
+
+    def test_training(self):
+        doc_ids = self.string_classifier.add_examples(self._NEW_EXAMPLES)
+        predicted_label = self.string_classifier.predict_label(doc_ids[0])
+        self.assertEquals(predicted_label, 'food')
+        predicted_label = self.string_classifier.predict_label(doc_ids[1])
+        self.assertEquals(predicted_label, 'pets')
+        predicted_label = self.string_classifier.predict_label(doc_ids[2], 0.7)
```

Comment:
If a default value is given, the arg name should be specified. In this case:

    self.string_classifier.predict_label(doc_ids[2], threshold=0.7)